### PR TITLE
fix: resolve BirdNET v3.0 ONNX outputs by name, not by position

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Every published BirdNET v3.0 model fails to load:

```
error: failed to build classifier: label count mismatch: model expects 1280, got 422
```

1280 is the embedding width, not a class count. The crate assumed v3.0 outputs are ordered `(embeddings, predictions)`; every published export has them the other way round, so detection read the embedding tensor's width as the class count and rejected the labels file that matched the real one.

## Why name-based, and why keyed on the embeddings output

Output order is not a property of a model family, and neither is the class-score name:

| model | outputs |
|---|---|
| BirdNET v3.0 fp32 | `predictions`, `embeddings` |
| BirdNET v3.0 fp16 (same weights) | `output`, `embeddings` |
| BirdNET v2.4 + embeddings | `output`, `model/GLOBAL_AVG_POOL/Mean_reduced_0` |
| Perch v2 | `embedding`, `spatial_embedding`, `spectrogram`, `label` |

Keying on the class-score name would have fixed one variant of v3.0 and left the other broken. The embeddings name is the stable one, so it identifies the pair and the other output is the class scores. A two-output model whose names say nothing falls back to class-scores-first, which is the layout of every published two-output model including v2.4-with-embeddings, and two outputs both claiming to be embeddings falls back rather than guessing.

## What changed

Roles are resolved once during detection and carried in `ModelConfig` as `predictions_index` and `embeddings_index`, so no consumer hardcodes an index. That covers the five places that each held their own copy of the assumption: both detection paths, the override path, single-segment inference, batch inference, and the `IoBinding` batch context. An explicit `--model v30` failed in exactly the same way as auto-detection, because the override path assumed the layout too.

Two further bugs fell out of the same reasoning:

- The dynamic-input path decided the family with "is the first output bigger", which is true for a v3.0 export listing 11,560 class scores ahead of 1,280-wide embeddings, so such a model was detected as v2.4 and run at 48 kHz with a 3 second window. Family now comes from the embedding width, which is what actually distinguishes the two.
- Perch's batch arm hardcoded outputs 0 and 3. It now uses the resolved indices like everything else, through a shared helper; the three batch arms differed on nothing but those indices.

## API change

`detect_model_type` takes an `output_names: &[String]` argument, and `ModelConfig` gains two fields. Both are breaking, which is why this targets the 2.0.0-rc line. Passing an empty name slice selects the documented positional fallback.

## Test Plan

- [x] `cargo fmt --check`, `cargo clippy --all-targets -- -D warnings`, `cargo test` all green
- [x] The two existing v3.0 detection tests asserted the wrong layout, which is why the bug looked correct for as long as it did. They now describe the published models.
- [x] New resolver tests: both namings, both orders, singular `embedding`, case insensitivity, the v2.4 opaque-name fallback, the both-claim-embeddings ambiguity, the empty-names fallback, and the override path
- [x] Verified against real models through birda, with a local path patch: the nordic v3.0 slice in **fp16** and **fp32** each produce 205 detections with *Glaucidium passerinum* top at 0.731 in 40 of 41 segments, which is the bird actually in the recording, and the two variants agree exactly
- [x] Perch v2 regional unaffected: same detections before and after

## Follow-up

Once this is tagged and published, birda needs its dependency bumped to `2.0.0-rc.16`; its gallery already ships v3.0 entries that install correctly but cannot run until then.

### CODE REVIEW

File: src/classifier.rs [hunk 1]
```diff
@@ -983,6 +1003,24 @@ impl Classifier {
     }
 
     /// Process batch inference outputs
+    /// Class-score and embedding tensors for a batch, by resolved output index.
+    ///
+    /// One place that knows which output is which, so a family's layout is
+    /// stated once at detection rather than repeated per arm. The arms differed
+    /// on nothing else: v2.4, v3.0 and Perch all read the same two tensors and
+    /// slice them identically.
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/classifier.rs:988]: Fixed in cf0c999. Correct, and it was the same mistake one layer down: detection resolves roles for v2.4-with-embeddings as well, so config and inference could disagree for a model whose names resolved to anything other than 0 and 1. Keeping a per-family index at each call site is precisely what let one wrong assumption live in five places, so there is now a single rule everywhere, including the BSG and Perch paths: ask the config. Verified against real models that v2.4, v2.4-with-embeddings, Perch regional and v3.0 regional all still produce the same detections, with the two v2.4 paths identical.

File: src/classifier.rs [hunk 2]
```diff
@@ -959,10 +974,15 @@ impl Classifier {
...
             }
             ModelType::BirdNetV30 => {
-                // Two outputs: embeddings at 0, predictions at 1
-                let embeddings = extract_tensor_data(outputs, 0)?;
-                let logits = extract_tensor_data(outputs, 1)?;
-                (Some(embeddings), logits)
+                // Indices resolved during detection, because output order is
+                // not a property of the family: the fp32 and fp16 exports of
+                // the same weights name their class scores differently.
+                let logits = extract_tensor_data(outputs, self.inner.config.predictions_index)?;
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/classifier.rs]: Fixed in cf0c999, same change as the Sentry finding on this file. Both v2.4 arms, single-segment and batch, plus the IoBinding context, now use `predictions_index` and `embeddings_index`. There is also a new test that walks every supported family and asserts each resolved index is in range and points at a tensor whose width matches what the config claims, since inference now reads whatever these point at.

File: src/detection.rs [hunk 3]
```diff
@@ -3,18 +3,79 @@
...
+        .map(|(index, _)| index)
+        .collect();
+
+    // Exactly one match, or the name told us nothing useful. Two outputs both
+    // claiming to be embeddings is a model we do not understand, and guessing
+    // between them would be worse than the documented default.
+    match embedding_positions.as_slice() {
+        [embeddings] if output_names.len() == 2 => OutputRoles {
+            predictions: 1 - embeddings,
+            embeddings: *embeddings,
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/detection.rs:57]: This one does not hold: `1 - embeddings` compiles. `impl Sub<&usize> for usize` exists in core, so subtracting a `&usize` from a `usize` is fine without dereferencing. The branch is also covered: `test_roles_find_the_embeddings_output_in_either_position` asserts `predictions: 1` for the embeddings-first layout, which only passes if that arithmetic runs. CI is green on Linux, macOS and Windows. Leaving it as is.

File: src/classifier.rs [hunk 4]
```diff
@@ -983,6 +980,24 @@ impl Classifier {
...
+            Error::Inference("embeddings index missing for a model with embeddings".into())
+        })?;
+        let logits = extract_tensor_data(outputs, self.inner.config.predictions_index)?;
+        let embeddings = extract_tensor_data(outputs, embeddings_index)?;
+        Ok((logits, embeddings))
+    }
+
     fn process_batch_outputs(
         &self,
         outputs: &ort::session::SessionOutputs,
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/classifier.rs:1003]: Fixed in a7e2ca4. You are right that it was safe today and wrong in principle: detection does resolve that index to 0 for v2.4 and BSG, but it contradicts the rule the rest of this change establishes, and it would go wrong silently the day that stopped being true. There is now no hardcoded output index left in the crate. Re-verified against real models: v3.0 regional 205 detections, Perch regional 41, v2.4 67, v2.4-with-embeddings 67, unchanged.

File: src/batch_context.rs [hunk 5]
```diff
@@ -280,20 +284,20 @@ impl BatchInferenceContext {
...
             }
             ModelType::BirdNetV30 => {
-                // Two outputs: embeddings and logits
-                let output_0 = self.output_names.first().ok_or_else(|| {
-                    Error::Inference("model missing first output tensor".to_string())
-                })?;
-                let output_1 = self.output_names.get(1).ok_or_else(|| {
-                    Error::Inference("model missing second output tensor".to_string())
-                })?;
+                // Bound by resolved role rather than by position. Binding both
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/batch_context.rs:286]: Fixed in fd0cf9a. That was the last positional output access in the crate; everything now goes through the resolved index or the name derived from it. Re-verified against real models: v3.0 regional 205 detections, Perch regional 41, v2.4 67, v2.4-with-embeddings 67, all unchanged.
